### PR TITLE
[quant] Dynamic quantization for bias.

### PR DESCRIPTION
**Pull request by dskhudia**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26057 [quant] Dynamic quantization for bias.**

bias is now unquantized (i.e. floating type) for qconv and qlinear. It is dynamically quantized by fbgemm.

~~TODO: Add some performance numbers.~~

Performance comparison on [Xeon Gold 6138 ](https://ark.intel.com/content/www/us/en/ark/products/120476/intel-xeon-gold-6138-processor-27-5m-cache-2-00-ghz.html). Single thread mode. It includes all unique conv shapes from Resnext101-32x4d. It got faster despite dynamic quantization of bias because we removed dequantize/quantize from kernel implementations.

% Change (With vs Without) (-ve change means faster)

Tests:

test:quantization
```
Summary (total time 8.41s):
  PASS: 24
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0More details at https://our.intern.facebook.com/intern/buck/build/74d5f6f7-55c9-4350-a618-2013042fffd8

  OMIT: 0
```

test:quantized
```
Summary (total time 13.21s):
  PASS: 43
  FAIL: 0
  SKIP: 5
    caffe2/test:quantized - test_qnnpack_maxpool2d (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_compare_tensor_scalar (test_quantized.TestComparatorOps)
    caffe2/test:quantized - test_qnnpack_linear (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_relu (test_quantized.TestQNNPackOps)
    caffe2/test:quantized - test_qnnpack_add (test_quantized.TestQNNPackOps)
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: [D17328028](https://our.internmc.facebook.com/intern/diff/D17328028/)

### CODE REVIEW

**Comment by jianyuh** [comment]
Saw the OSS failure:
```
Sep 11 22:19:16 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:195:39: error: no matching function for call to 'fbgemm::ReQuantizeOutput<true, (fbgemm::QuantizationGranularity)2, float, int, fbgemm::DoNothing<float, float> >::ReQuantizeOutput(fbgemm::DoNothing<unsigned char, unsigned char>&, float*, int32_t&, int32_t&, int*, int32_t*, int*, const float*&, int64_t&, int, float*)'
Sep 11 22:19:16                act_times_w_scale.data());
Sep 11 22:19:16                                        ^
```
Is it because the FBGEMM code changes are not propagated to PyTorch?

**Comment by dskhudia** [comment]
> Saw the OSS failure:
> 
> ```
> Sep 11 22:19:16 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:195:39: error: no matching function for call to 'fbgemm::ReQuantizeOutput<true, (fbgemm::QuantizationGranularity)2, float, int, fbgemm::DoNothing<float, float> >::ReQuantizeOutput(fbgemm::DoNothing<unsigned char, unsigned char>&, float*, int32_t&, int32_t&, int*, int32_t*, int*, const float*&, int64_t&, int, float*)'
> Sep 11 22:19:16                act_times_w_scale.data());
> Sep 11 22:19:16                                        ^
> ```
> 
> Is it because the FBGEMM code changes are not propagated to PyTorch?

Yes. Once the fbgemm diffs land, it should work fine.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 1]
```diff
@@ -165,16 +142,22 @@ class QLinearInt8 final : public torch::OperatorKernel {
...
-          /*row_offsets=*/packA.getRowOffsetBuffer(),
-          /*col_offsets=*/col_offsets.data(),
-          /*bias=*/bias_ptr,
-          /*nCol=*/N);
+      fbgemm::ReQuantizeOutput<
+          ReluFused,
+          fbgemm::QuantizationGranularity::TENSOR,
+          float>
+          outputProcObj(
+              doNothingObj,
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/qlinear.cpp]: Maybe keep the comments for better annotating these arguments.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 2]
```diff
@@ -196,17 +179,20 @@ class QLinearInt8 final : public torch::OperatorKernel {
...
-              /*nextop=*/doNothingObj,
-              /*C_multiplier=*/output_multiplier_float.data(),
-              /*C_zero_point=*/output_zero_point_int32,
-              /*Aq_zero_point=*/input_zero_point_int32,
-              /*Bq_zero_point=*/pack_ptr.w_zp.data(),
-              /*row_offsets=*/packA.getRowOffsetBuffer(),
-              /*col_offsets=*/col_offsets.data(),
-              /*bias=*/bias_ptr,
-              /*nCol=*/N);
+              doNothingObj,
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/qlinear.cpp]: Ditto.

File: torch/nn/quantized/modules/linear.py [hunk 3]
```diff
@@ -229,16 +229,9 @@ def from_float(cls, mod):
         act_scale, act_zp = activation_observer.calculate_qparams()
         assert weight_observer.dtype == torch.qint8, 'Weight observer must have dtype torch.qint8'
         wt_scale, wt_zp = weight_observer.calculate_qparams()
-        # Scale bias to activation_scale/2^16, this quantizes bias
-        # to about 24 bits of precision
-        bias_scale = float(act_scale / (2**16))
```
Comment (jianyuh) [inline torch/nn/quantized/modules/linear.py]: The bias handler logic in the new FBGEMM API is different from "activation_scale/2^16" here. Will this cause any numerical diffenrece?

Comment (jianyuh) [reply torch/nn/quantized/modules/linear.py]: Just saw your comment that the bias is just not quantized, so I guess the accuracy should be improved with the new API.

Comment (raghuramank100) [reply torch/nn/quantized/modules/linear.py]: This was a temporary change, any numeric change will be below 1 LSB as bias was previously quantized to ~24 bit precision.

**Comment by raghuramank100** [comment]
Looks good, have you also checked the output of the  test: caffe2/test/test_quantized_models.py to ensure that overall numerics at the model level stay close?

**Review by raghuramank100 — APPROVED** [review]
Looks good, please check end to end model numerics before submitting.

**Comment by dskhudia** [comment]
> Looks good, have you also checked the output of the test: caffe2/test/test_quantized_models.py to ensure that overall numerics at the model level stay close?

Yes. That test runs as a part of test:quantized and the results for this are shown in the description above.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 1 above)
Comment (dskhudia) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: I removed the trivial ones and kept ones which provide some context.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 2 above)
Comment (dskhudia) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: Same reply as above.

**Review by jianyuh — APPROVED** [review]
LGTM!

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 4]
```diff
@@ -133,53 +133,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+      bias = bias.contiguous();
+      TORCH_CHECK(bias.dim() == 1, "bias should be a vector (1D Tensor)");
       TORCH_CHECK(
-          qbias.size(0) == K,
+          bias.size(0) == K,
           "bias should have K elements: " + std::to_string(K));
-      auto bias_contig = qbias.contiguous();
-      bias_ptr =
-          reinterpret_cast<int32_t*>(bias_contig.data_ptr<c10::qint32>());
+      bias_ptr = reinterpret_cast<float*>(bias.data_ptr<float>());
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: you don't need reinterpret_cast if it's float

File: torch/nn/quantized/functional.py [hunk 5]
```diff
@@ -77,7 +77,7 @@ def conv2d(input, weight, bias,
     Args:
         input: quantized input tensor of shape :math:`(\text{minibatch} , \text{in\_channels} , iH , iW)`
         weight: quantized filters of shape :math:`(\text{out\_channels} , \frac{\text{in\_channels}}{\text{groups}} , kH , kW)`
-        bias: **non-quantized** bias tensor of shape :math:`(\text{out\_channels})`. The tensor type must be `torch.int32`.
+        bias: **non-quantized** bias tensor of shape :math:`(\text{out\_channels})`. The tensor type must be `torch.float`.
```
Comment (dzhulgakov) [inline torch/nn/quantized/functional.py]: @supriyar - is qnnpack also expecting fp32 bias? I think it's best to keep them the same. (fp32 itself sounds fine, we can re-quantize inside pack)

If they are different, given that quantization of bias is cheap, we could allow this function to receive either quantized or non-quantized bias and do conversion internally if needed.

**Review by dzhulgakov — COMMENTED** [review]
@supriyar - Let's make sure that it's qnnpack-compatible

File: torch/nn/quantized/functional.py (hunk 5 above)
Comment (supriyar) [reply torch/nn/quantized/functional.py]: qnnpack expects int32 bias. We prepack weight and bias together in qnnpack 
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/pack.h#L81

We decided to not change qnnpack (in the interest of time) to decouple this and instead pack bias in the fbgemm struct to make them semantically the same.
Given that I think 2nd option proposed by @dzhulgakov makes sense - i.e. change the function to accept both quantized and non-quantized bias.

File: aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp [hunk 6]
```diff
@@ -48,7 +48,7 @@ class QLinearPackWeightInt8 final : public c10::OperatorKernel {
   at::Tensor operator()(at::Tensor weight, c10::optional<Tensor> bias) {
     TORCH_CHECK(
         weight.dim() == 2,
-        "The weight tensor for quantized::linear_prepack should"
+        "The weight tensor for quantized::linear_prepack (fbgemm) should"
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp]: revert? - it's not a specific requirement for fbgemm :)

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 7]
```diff
@@ -133,53 +133,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
+      bias = bias.contiguous();
+      TORCH_CHECK(bias.dim() == 1, "bias should be a vector (1D Tensor)");
       TORCH_CHECK(
-          qbias.size(0) == K,
+          bias.size(0) == K,
           "bias should have K elements: " + std::to_string(K));
-      auto bias_contig = qbias.contiguous();
-      bias_ptr =
-          reinterpret_cast<int32_t*>(bias_contig.data_ptr<c10::qint32>());
+      bias_ptr = bias.data_ptr<float>();
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: I'd add a nice TORCH_CHECK that the input is unquantized with a human-understandable message. data_ptr would do one for you but it's nicer to call out that we expect bias in fp32

File: aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp (hunk 6 above)
Comment (dskhudia) [reply aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp]: Not a part of this diff now.

File: aten/src/ATen/native/quantized/cpu/qconv.cpp (hunk 7 above)
Comment (dskhudia) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Added a check.

**Pull request closed without merging**